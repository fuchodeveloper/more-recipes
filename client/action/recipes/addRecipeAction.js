import { batchActions } from 'redux-batched-actions';
import axios from 'axios';
import alertify from 'alertify.js';
import { ADD_RECIPE, ADD_RECIPE_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description create a recipe
 *
 * @export addRecipeActionCreator
 *
 * @param {Object} recipe
 *
 * @returns {Object} recipe
 */
const addRecipeActionCreator = recipe => ({
  type: ADD_RECIPE,
  recipe
});

/**
 * @description create a recipe error
 *
 * @export addRecipeActionError
 *
 * @param {Object} error
 *
 * @returns {Object} error
 */
const addRecipeActionError = error => ({
  type: ADD_RECIPE_ERROR,
  error
});

const addRecipeAction = recipe => (dispatch) => {
  dispatch(setFetching());
  return axios.post('/api/v1/recipes', recipe)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(batchActions([
        dispatch(addRecipeActionCreator(response.data.recipe)),
        unsetFetching()
      ]));
    })
    .catch((error) => {
      dispatch([
        dispatch(addRecipeActionError(error.response.data.error)),
        unsetFetching()
      ]);
    });
};

export default addRecipeAction;
