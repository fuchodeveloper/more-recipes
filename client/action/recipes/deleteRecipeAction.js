import axios from 'axios';
import alertify from 'alertify.js';
import { DELETE_RECEIPE, DELETE_RECEIPE_FAIL } from '../types';
import networkError from '../networkError';

export const deleteRecipeActionError = error => ({
  type: DELETE_RECEIPE_FAIL,
  error
});

export const deleteRecipeActionCreator = recipes => ({
  type: DELETE_RECEIPE,
  recipes
});

const deleteRecipeAction = recipeId => dispatch =>
  axios.delete(`/api/v1/recipes/${recipeId}`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(deleteRecipeActionCreator(response.data.recipes));
    })
    .catch((error) => {
      if (!error.response) {
        return networkError(error);
      }
      dispatch(deleteRecipeActionError(error.response.data));
    });

export default deleteRecipeAction;
