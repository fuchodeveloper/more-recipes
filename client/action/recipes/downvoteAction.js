import axios from 'axios';
import alertify from 'alertify.js';
import { GET_DOWNVOTE } from '../types';

/**
 * Recipe doenvote action creator
 * @param {recipe} recipe
 * @returns {recipe} recipe
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: GET_DOWNVOTE,
  recipe
});

/**
 * Authenticated user can downvote a recipe
 *
 * @export func
 * @param {integer} param
 * @returns {object} object
 */
const downvoteRecipe = param =>
  dispatch => axios.post(`/api/v1/recipes/${param}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      alertify.logPosition('bottom right');
      alertify.error(error.message);
    });


export default downvoteRecipe;
