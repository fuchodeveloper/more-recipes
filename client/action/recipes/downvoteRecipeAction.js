import axios from 'axios';
import alertify from 'alertify.js';
import { DOWNVOTE_RECIPE, DOWNVOTE_RECIPE_ERROR } from '../types';
import networkError from '../networkError';

/**
 * @description Recipe downvote action creator
 *
 * @param {recipe} recipe recipe parameter
 *
 * @returns {recipe} recipe downvoted
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: DOWNVOTE_RECIPE,
  recipe
});
/**
 * @description Recipe downvote error
 *
 * @param {Object} error error parameter
 *
 * @returns {Object} error returns the error object
 *
 */
const downvoteRecipeActionError = error => ({
  type: DOWNVOTE_RECIPE_ERROR,
  error
});

/**
 * @description Authenticated user can downvote a recipe
 *
 * @export downvoteRecipe
 *
 * @param {Number} id
 *
 * @returns {object} downvoted recipe object
 */
const downvoteRecipeAction = id =>
  dispatch => axios.post(`/api/v1/recipes/${id}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.recipe.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      if (!error.response) {
        return networkError(error);
      }
      alertify.delay(1000);
      alertify.logPosition('bottom right');
      alertify.error('Please login to downvote recipe');
      dispatch(downvoteRecipeActionError(error.response.data));
    });


export default downvoteRecipeAction;
