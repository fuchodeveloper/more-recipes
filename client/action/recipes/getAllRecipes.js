import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import { GET_ALL_RECIPES, GET_PAGE_DETAILS } from '../types';
import { setFetching, unsetFetching } from '../fetching';

export const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

export const recipePageDetails = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipes = page => (dispatch) => {
  dispatch(setFetching());
  axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        receiveRecipes(response.data.recipes),
        recipePageDetails(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((error) => {
      // dispatch(batchActions([

      // ]));
      console.log(error.message);
    });
};

export default getAllRecipes;
