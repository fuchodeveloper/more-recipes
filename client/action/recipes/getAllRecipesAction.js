import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import
{ GET_ALL_RECIPES, GET_PAGE_DETAILS, GET_ALL_RECIPES_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description get all recipes action creator
 *
 * @param {Object} recipes recipes object parameter
 *
 * @returns {Object} recipes returns all recipes object
 */
const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

/**
 * @description get all recipes error action
 *
 * @param {Object} error recipes error object
 *
 * @returns {Object} error returns get all recipes error object
 */
const receiveRecipesError = error => ({
  type: GET_ALL_RECIPES_ERROR,
  error
});

/**
 * @description handles page detail
 *
 * @param {Number} pageCount recipe page count parameter
 *
 * @returns {Number} returns recipes page Count
 */
const recipePageCount = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipesAction = page => (dispatch) => {
  dispatch(setFetching());
  return axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        dispatch(receiveRecipes(response.data.recipes)),
        recipePageCount(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((serverError) => {
      dispatch(receiveRecipesError(serverError.recipes.data.error));
    });
};

export default getAllRecipesAction;
