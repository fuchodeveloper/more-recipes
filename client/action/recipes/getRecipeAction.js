import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import { RECEIVE_RECIPE, RECEIVE_RECIPE_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * Set type for GET all recipes
 *
 * @export {function} export function receiveRecipe
 *
 * @param {Object} recipe recipe object parameter
 * @param {Object} favorited favorited object parameter
 *
 * @returns {Object} returns requested recipe
 * @returns {Object} returned favorited
 */
export const receiveRecipe = (recipe, favorited) => ({
  type: RECEIVE_RECIPE,
  recipe,
  favorited
});

/**
 * @description recieve recipe error action creator
 *
 * @param {Object} error error object parameter
 *
 * @returns {Object} error recipe error
 */
const receiveRecipeError = error => ({
  type: RECEIVE_RECIPE_ERROR,
  error
});

/**
 * GET a recipe
 * @param {Number} id
 *
 * @export {function} getRecipeAction
 *
 * @returns {Object} dispatch - dispatch the get recipe action
 */
const getRecipeAction = id => (dispatch) => {
  dispatch(setFetching());
  return axios.get(`/api/v1/recipes/${id}`)
    .then((response) => {
      const recipeObject = {
        recipe: response.data.recipe,
        favorited: response.data.favorited
      };
      dispatch(batchActions([
        dispatch(receiveRecipe(recipeObject)),
        unsetFetching()
      ]));
    })
    .catch((error) => {
      dispatch(receiveRecipeError(error.response.data.error));
    });
};
export default getRecipeAction;
