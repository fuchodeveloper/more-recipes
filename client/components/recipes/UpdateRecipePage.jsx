/* eslint-disable no-undef */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import getRecipeAction from '../../action/recipes/getRecipeAction';
import updateRecipeAction from '../../action/recipes/updateRecipeAction';
import validateRecipe from '../../validations/validateRecipe';

/**
 * @description component to handle recipe update
 *
 * @class UpdateRecipePage
 *
 * @extends {React.Component}
 */
class UpdateRecipePage extends React.Component {
  /**
   * @description Creates an instance of UpdateRecipePage.
   *
   * @param {Object} props
   *
   * @memberof UpdateRecipePage
   */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      image: '',
      ingredients: '',
      direction: '',
      errors: {
        name: '',
        ingredients: '',
        direction: '',
        image: ''
      },
      redirect: false
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.uploadWidget = this.uploadWidget.bind(this);
  }

  /**
 * @description component lifecycle method called on component mount
 *
 * @memberof UpdateRecipePage
 *
 * @returns {undefined}
 */
  componentWillMount() {
    const { id } = this.props.match.params;
    this.props.getRecipe(id);
  }

  /**
 * @description component lifecycle method
 *
 * @param {Object} nextProps
 *
 * @memberof UpdateRecipePage
 *
 * @returns {undefined}
 */
  componentWillReceiveProps(nextProps) {
    const {
      name, image, ingredients, direction
    } = nextProps.recipe;
    this.setState({
      name, image, ingredients, direction
    });
  }

  /**
   * @description Handle change events
   *
   * @param {event} event
   *
   * @returns {undefined}
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
 * @description function to handle form submission
 *
 * @param {Object} event
 *
 * @memberof UpdateRecipePage
 *
 * @returns {undefined}
 */
  onSubmit(event) {
    event.preventDefault();

    /**
     * @description Id of the recipe to be updated sent as param
     */
    const { id } = this.props.match.params;

    if (this.isValid()) {
      this.setState({ errors: {} });
      this.props.updateRecipe(id, this.state)
        .then(() => {
          this.setState({ redirect: true });
        });
    }
  }

  /**
   * @description Handle form validation
   *
   * @returns {Object} isValid
   *
   * @memberof UpdateRecipePage
   */
  isValid() {
    const { errors, isValid } = validateRecipe(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
 * @description widget function to handle image upload
 *
 * @param {Object} event
 *
 * @memberof UpdateRecipePage
 *
 * @returns {undefined}
 */
  uploadWidget(event) {
    event.preventDefault();
    const scope = this;

    cloudinary.openUploadWidget(
      {
        cloud_name: 'fuchodeveloper',
        upload_preset: 'wvxnziq0',
        tags: ['recipe']
      },
      (error, result) => {
        scope.setState({ image: result[0].url });
      }
    );
  }

  /**
 * @description renders component to DOM
 *
 * @memberof UPdateRecipePage
 *
 *  @returns {JSX} JSX representation of component
 */
  render() {
    if (this.state.redirect) {
      const { id } = this.props.match.params;
      return <Redirect to={`/recipes/${id}`} />;
    }

    const {
      name, image, direction, ingredients, isLoading
    } = this.props.recipe;

    if (isLoading) {
      return (
        <h2 className="text-center">Loading...</h2>
      );
    }

    return (
      <div>

        <div className="container margin-top-70">
          <div>
            <h1 className="text-center p-4 center-hero-text">
            Update Recipe
            </h1>
          </div>

          <div className="col-md-6 mx-auto p-3">
            {/* Recipe avatar */}
            <div className="col-md-6 mx-auto p-3 add_recipe-card text-center">
              <img
                src={this.state.image}
                width="350px"
                id="recipe-image-avatar"
                height="auto"
                className="img-fluid"
                alt={this.state.name}
              />
            </div>

            <form encType="multipart/form-data" onSubmit={this.onSubmit}>
              <br />
              <div className="upload text-center">
                <button
                  name="image"
                  id="image"
                  onClick={this.uploadWidget}
                  className="upload-button btn-primary btn-primary-color"
                >
                  Add Image
                </button>
              </div>

              <h3 className="text-center p-4 center-hero-text">
              Add Ingredients
              </h3>

              <div className="form-row" id="recipe-quantity">
                <div className="col-sm-12">
                  <div>
                    <p className="text-muted">Recipe Name</p>
                    <input
                      placeholder="Enter recipe name"
                      type="text"
                      name="name"
                      id="name"
                      className="form-control"
                      value={this.state.name}
                      onChange={this.onChange}
                      required
                    />
                    { this.state.errors.name &&
                    <div className="text-danger">
                      {this.state.errors.name}
                    </div>
                    }
                  </div>
                </div>
                <div className="m-1" />

                <div className="col-sm-12">
                  <div>
                    <br />
                    <p className="text-muted">Recipe Ingredients</p>
                    <textarea
                      placeholder="Enter recipe ingredients"
                      cols="30"
                      rows="5"
                      name="ingredients"
                      id="ingredient"
                      className="form-control"
                      value={this.state.ingredients}
                      onChange={this.onChange}
                      required
                    />
                    { this.state.errors.ingredients &&
                      <div className="text-danger">
                        {this.state.errors.ingredients}
                      </div>
                    }
                  </div>
                </div>
              </div>

              <div className="clearfix" />
              <div className="form-row">
                <div className="col-md-12 mb-3">
                  <label htmlFor="direction" />
                  <p className="text-muted">Recipe Direction</p>
                  <textarea
                    className="form-control"
                    placeholder="Give a detailed direction for this recipe:"
                    name="direction"
                    id="recipe-direction"
                    cols="30"
                    rows="10"
                    value={this.state.direction}
                    onChange={this.onChange}
                    required
                  />
                  { this.state.errors.direction &&
                  <div className="text-danger">
                    {this.state.errors.direction}
                  </div>
                    }
                </div>
              </div>

              <div className="float-right p-1">
                <input
                  type="submit"
                  className="btn btn-primary btn-primary-color"
                  value="Submit"
                />

              </div>

            </form>
          </div>
        </div>
        <div className="clearfix m-5" />

      </div>

    );
  }
}

UpdateRecipePage.propTypes = {
  getRecipe: PropTypes.func.isRequired,
  updateRecipe: PropTypes.func.isRequired
};

UpdateRecipePage.contextTypes = {
  router: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  recipe: state.recipesReducer
});
const mapDispatchToProps = dispatch => ({
  updateRecipe: (recipeId, recipe) =>
    dispatch(updateRecipeAction(recipeId, recipe)),
  getRecipe: recipeId => dispatch(getRecipeAction(recipeId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UpdateRecipePage);
