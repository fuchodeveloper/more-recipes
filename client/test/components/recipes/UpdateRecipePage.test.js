import React from 'react';
import { shallow } from 'enzyme';
import expect from 'expect';
import sinon from 'sinon';
import { UpdateRecipePage } from '../../../components/recipes/UpdateRecipePage';
import mockData from '../../__mocks__/mockData';

let props;

const setup = () => {
  props = {
    cloudinary: {
      openUploadWidget: jest.fn(() => Promise.resolve())
    },
    updateRecipe: jest.fn(() => Promise.resolve()),
    match: {
      params: {
        id: 1
      }
    },
    recipe: {
      name: 'jollof rice',
      image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/'
      + 'v1516760699/noodles_c6ltkq.jpg',
    },
    getRecipe: jest.fn()
  };
  return shallow(<UpdateRecipePage {...props} />);
};

describe('UpdateRecipePage component', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should set the recipe name when input value changes', () => {
    const event = {
      target: { name: 'name', value: '' }
    };
    const wrapper = setup();
    const recipeName = wrapper.find('#name');

    event.target.value = 'jollof rice';
    recipeName.simulate('change', event);

    expect(wrapper.instance().state.name).toBe('jollof rice');
  });

  it('should set the recipe ingredients when input value changes', () => {
    const event = {
      target: { name: 'ingredients', value: '' }
    };
    const wrapper = setup();
    const recipeIngredients = wrapper.find('#ingredients');

    event.target.value = 'rice and maggi';
    recipeIngredients.simulate('change', event);

    expect(wrapper.instance().state.ingredients).toBe('rice and maggi');
  });

  it('should simulate click event on image button to call uploadWidget', () => {
    const event = {
      preventDefault: jest.fn()
    };

    const wrapper = setup();
    const recipeImage = wrapper.find('#image-button');

    recipeImage.simulate('click', event);

    expect(wrapper.instance().uploadWidget(event)).toMatchSnapshot();
  });

  it('should set the recipe image on click and call uploadWidget', () => {
    const event = {
      preventDefault: jest.fn()
    };

    const wrapper = setup();
    const recipeImage = wrapper.find('#image-button');

    recipeImage.simulate('click', event);
    wrapper.setState({
      image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/'
      + 'v1516760699/noodles_c6ltkq.jpg'
    });
    expect(wrapper.instance().uploadWidget(event)).toMatchSnapshot();
  });

  it('should not update recipe if state is empty', () => {
    const event = {
      preventDefault: jest.fn()
    };
    const wrapper = setup();
    const form = wrapper.find('#update-recipe-forms');

    form.simulate('submit', event);
  });

  it('should update recipe if state is set currently', () => {
    const event = {
      preventDefault: jest.fn()
    };
    const { updateRecipe } = mockData;
    const wrapper = setup();
    const form = wrapper.find('#update-recipe-forms');
    wrapper.setState(updateRecipe);

    expect(wrapper.instance().state.name).toEqual(updateRecipe.name);
    form.simulate('submit', event);
  });

  it('component should receive props', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const spy = sinon.spy(
      UpdateRecipePage.prototype,
      'componentWillReceiveProps'
    );
    const nextProps = {
      recipe: {
        recipe: {
          name: 'party jollof rice and beans',
          image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/'
          + 'v1516760699/noodles_c6ltkq.jpg',
          ingredients: 'local rice and beans',
          direction: 'proin eget tortor risus. nulla quis lorem ut libero '
          + 'malesuada feugiat. curabitur non nulla  sit amet nisl tempus '
          + 'convallis quis ac lectus. curabitur arcu erat,  accumsan id '
          + 'imperdiet et, porttitor at sem. cras ultricies ligula sed magna '
          + 'dictum porta.',
        }
      }
    };
    shallow(<UpdateRecipePage {...props} componentWillReceiveProps={spy} />);
    action.componentWillReceiveProps(nextProps);

    expect(action.state.name).toBe(nextProps.recipe.recipe.name);
    expect(action.state.ingredients).toBe(nextProps.recipe.recipe.ingredients);
  });
});
