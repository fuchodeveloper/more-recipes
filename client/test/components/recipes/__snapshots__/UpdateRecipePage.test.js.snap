// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateRecipePage component should render with right amount of elements 1`] = `undefined`;

exports[`UpdateRecipePage component should set the recipe image on click and call uploadWidget 1`] = `undefined`;

exports[`UpdateRecipePage component should simulate click event on image button to call uploadWidget 1`] = `undefined`;
